### PR TITLE
Use azure subscription as source credentials for backup

### PR DESCRIPTION
**Pull request by swapnilgm**
Signed-off-by: Swapnil Mhamane <swapnil.mhamane@sap.com>

**What this PR does / why we need it**:
This PR modifies azure backup extension controller to interpret backup bucket referred secret to have subscription details.
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator

```

### CODE REVIEW

File: controllers/provider-azure/pkg/azure/client/storage.go [hunk 1]
```diff
@@ -19,17 +19,93 @@ import (
...
 	"github.com/gardener/gardener-extensions/controllers/provider-azure/pkg/azure"
+	"github.com/gardener/gardener-extensions/controllers/provider-azure/pkg/internal"
 	extensionscontroller "github.com/gardener/gardener-extensions/pkg/controller"
 
 	corev1 "k8s.io/api/core/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+// NewStorageClientAuthFromSubsriptionSecretRef retrieves the azure storage client auth from specified by the secret reference.
+func NewStorageClientAuthFromSubsriptionSecretRef(ctx context.Context, c client.Client, secretRef *corev1.SecretReference, resourceGroupName, accountName, region string) (*StorageAuth, error) {
```
Comment (rfranzke) [inline controllers/provider-azure/pkg/azure/client/storage.go]:
```suggestion
func NewStorageClientAuthFromSubscriptionSecretRef(ctx context.Context, c client.Client, secretRef *corev1.SecretReference, resourceGroupName, accountName, region string) (*StorageAuth, error) {
```

File: controllers/provider-azure/pkg/azure/client/storage.go [hunk 2]
```diff
@@ -19,17 +19,93 @@ import (
...
+		Kind:     storage.BlobStorage,
+		Location: &region,
+		AccountPropertiesCreateParameters: &storage.AccountPropertiesCreateParameters{
+			AccessTier: storage.Cool,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	if err := future.WaitForCompletionRef(ctx, storageAccountClient.Client); err != nil {
```
Comment (rfranzke) [inline controllers/provider-azure/pkg/azure/client/storage.go]:
```suggestion

	if err := future.WaitForCompletionRef(ctx, storageAccountClient.Client); err != nil {
```

File: controllers/provider-azure/pkg/azure/client/storage.go [hunk 3]
```diff
@@ -19,17 +19,93 @@ import (
...
+		return nil, err
+	}
+
+	keysResponse, err := storageAccountClient.ListKeys(ctx, resourceGroupName, accountName)
+	if err != nil {
+		return nil, err
+	}
+
+	key := (*keysResponse.Keys)[0]
```
Comment (rfranzke) [inline controllers/provider-azure/pkg/azure/client/storage.go]:
```suggestion
	key := keysResponse.Keys[0]
```

However, should be checked whether `keys != nil && len(keys) > 0`?

Comment (swapnilgm) [reply controllers/provider-azure/pkg/azure/client/storage.go]: There will be key associated with StorageAccount. So don't have to check that. The suggestion will lead to sysntex error. Its pointer to the array.

Comment (rfranzke) [reply controllers/provider-azure/pkg/azure/client/storage.go]: Ah,ok, thanks. Alright!

File: controllers/provider-azure/pkg/controller/backupbucket/actuator.go [hunk 4]
```diff
@@ -53,10 +59,70 @@ func (a *actuator) Reconcile(ctx context.Context, bb *extensionsv1alpha1.BackupB
...
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	bb.Status.GeneratedSecretRef = &corev1.SecretReference{
+		Name:      generatedSecret.Name,
+		Namespace: generatedSecret.Namespace,
+	}
```
Comment (rfranzke) [inline controllers/provider-azure/pkg/controller/backupbucket/actuator.go]: call update status here

Comment (swapnilgm) [reply controllers/provider-azure/pkg/controller/backupbucket/actuator.go]: Done

**Pull request merged**